File: minsoo/baekjoon/15684.cc
// https://www.acmicpc.net/problem/15684

#include <cstdio>

int ROW, COL, NUM_CONN, minAdd = 4;
int ladder[311]{};

bool isColIdentity(int cIn) {
    int cOut = cIn;
    for (int r = 1; r <= ROW; r++) {
        if (ladder[10 * r + cOut])
            cOut++;
        else if (ladder[10 * r + cOut - 1])
            cOut--;
    }
    return cOut == cIn;
}

bool isIdentity() {
    for (int cIn = 1; cIn <= COL; cIn++)
        if (!isColIdentity(cIn))
            return false;
    return true;
}

void dfs(int begin, int end, int nAdded, int nAdd) {
    if (minAdd < 4)
        return;
    if (nAdded == nAdd) {
        if (isIdentity())
            minAdd = nAdded;
        return;
    }
    if (begin == end)
        return;

    for (int i = begin; i < end; i++) {
        if (ladder[i - 1] || ladder[i] || ladder[i + 1])
            continue;

        ladder[i] = 1;
        dfs(i + 1, end, nAdded + 1, nAdd);
        ladder[i] = 0;
    }
}

int main() {
    scanf("%d %d %d", &COL, &NUM_CONN, &ROW);
    for (int n = 0, r, c; n < NUM_CONN; n++) {
        scanf("%d %d", &r, &c);
        ladder[10 * r + c] = 1;
    }
    for (int nAdd = 0; nAdd < 4; nAdd++) {
        dfs(11, 10 * ROW + COL, 0, nAdd);
        if (minAdd < 4)
            break;
    }
    printf("%d\n", minAdd < 4 ? minAdd : -1);
}

File: minsoo/baekjoon/15684.py
# https://www.acmicpc.net/problem/15684


from dataclasses import dataclass, field


@dataclass
class LadderGame:
    r: int
    c: int
    connections: list[list[int, int]]
    ladder: list[bool] = field(init=False)

    def __post_init__(self):
        ladder = [False for _ in range(10 * (self.r + 1) + self.c + 1)]
        for r, c in self.connections:
            ladder[10 * r + c] = True
        self.ladder = ladder

    def ends_if_starts_from(self, c: int):
        for r in range(1, self.r + 1):
            if self.ladder[10 * r + c]:
                c += 1
            elif self.ladder[10 * r + c - 1]:
                c -= 1
        return c

    def is_identity(self):
        for c in range(1, self.c + 1):
            if self.ends_if_starts_from(c) != c:
                return False
        return True

    def can_be_identity(self, start: int = 11, num_connections_added: int = 0):
        if self.is_identity():
            return num_connections_added
        if start == 10 * self.r + self.c:
            return -1
        if num_connections_added == 3:
            return -1

        min_added = 4
        for connection in range(start, 10 * self.r + self.c):
            if (
                self.ladder[connection - 1]
                or self.ladder[connection]
                or self.ladder[connection + 1]
            ):
                continue

            self.ladder[connection] = True
            _num_connections_added = self.can_be_identity(
                start=connection + 1,
                num_connections_added=num_connections_added + 1,
            )
            self.ladder[connection] = False

            if _num_connections_added > -1:
                min_added = min(min_added, _num_connections_added)

        return -1 if min_added == 4 else min_added


c, n, r = map(int, input().split())
connections = [list(map(int, input().split())) for _ in range(n)]
ladder_game = LadderGame(r=r, c=c, connections=connections)

print(ladder_game.can_be_identity())
